crates/bevy_render/src/view/visibility/range.rs

```diff
@@ -0,0 +1,436 @@
+                    visibility_range.end_margin.end,
+                )) as u16)
+                .unwrap_or_default()
+            });
+
+        self.entities.insert(
+            entity,
+            RenderVisibilityEntityInfo {
+                buffer_index,
+                is_abrupt: visibility_range.is_abrupt(),
+            },
+        );
+    }
+
+    /// Returns the index in the GPU buffer corresponding to the visible range
+    /// for the given entity.
+    ///
+    /// If the entity has no visible range, returns `None`.
+    #[inline]
+    pub fn lod_index_for_entity(&self, entity: Entity) -> Option<NonMaxU16> {
+        self.entities.get(&entity).map(|info| info.buffer_index)
+    }
+
+    /// Returns true if the entity has a visibility range and it isn't abrupt:
+    /// i.e. if it has a crossfade.
+    #[inline]
+    pub fn entity_has_crossfading_visibility_ranges(&self, entity: Entity) -> bool {
+        self.entities
+            .get(&entity)
+            .is_some_and(|info| !info.is_abrupt)
+    }
+
+    /// Returns a reference to the GPU buffer that stores visibility ranges.
+    #[inline]
+    pub fn buffer(&self) -> &BufferVec<Vec4> {
+        &self.buffer
+    }
+}
+
+/// Stores which entities are in within the [`VisibilityRange`]s of views.
+///
+/// This doesn't store the results of frustum or occlusion culling; use
+/// [`super::ViewVisibility`] for that. Thus entities in this list may not
+/// actually be visible.
+///
+/// For efficiency, these tables only store entities that have
+/// [`VisibilityRange`] components. Entities without such a component won't be
+/// in these tables at all.
+///
+/// The table is indexed by entity and stores a 32-bit bitmask with one bit for
+/// each camera, where a 0 bit corresponds to "out of range" and a 1 bit
+/// corresponds to "in range". Hence it's limited to storing information for 32
+/// views.
+#[derive(Resource, Default)]
+pub struct VisibleEntityRanges {
+    /// Stores which bit index each view corresponds to.
+    views: EntityHashMap<Entity, u8>,
+
+    /// Stores a bitmask in which each view has a single bit.
+    ///
+    /// A 0 bit for a view corresponds to "out of range"; a 1 bit corresponds to
+    /// "in range".
+    entities: EntityHashMap<Entity, u32>,
+}
+
+impl VisibleEntityRanges {
+    /// Clears out the [`VisibleEntityRanges`] in preparation for a new frame.
+    fn clear(&mut self) {
+        self.views.clear();
+        self.entities.clear();
+    }
+
+    /// Returns true if the entity is in range of the given camera.
+    ///
+    /// This only checks [`VisibilityRange`]s and doesn't perform any frustum or
+    /// occlusion culling. Thus the entity might not *actually* be visible.
+    ///
+    /// The entity is assumed to have a [`VisibilityRange`] component. If the
+    /// entity doesn't have that component, this method will return false.
+    #[inline]
+    pub fn entity_is_in_range_of_view(&self, entity: Entity, view: Entity) -> bool {
```

Comment:
This code, to my knowledge, doesn't short circuit if `entity` doesn't exist in `self.entities`. Given that this is a high-frequency function (and the number of entities without ranges might be high), I think it makes sense to optimize this case to save the extra `view` hash in those cases. I'll push the change.

Reply:
Just pushed this change